### PR TITLE
Add option to API to update PullRequest base branch

### LINKED ISSUE #11552: Enable Update PR base branch through API
https://github.com/go-gitea/gitea/blob/88fe7b5a720365e5c8f0a347730f895263465311/routers/api/v1/repo/pull.go#L351

This API method, together with it's implementation in https://gitea.com/gitea/go-sdk should implement mutating the `base` branch just as available on the UI, based on adding this field into the following struct
https://github.com/go-gitea/gitea/blob/88fe7b5a720365e5c8f0a347730f895263465311/modules/structs/pull.go#L82-L93

This is required for full git-town support https://github.com/git-town/git-town/issues/1451 since git town extensively uses branch hierarchy: if the an ancestor get's merged, all it's children branches must be reassigned to the ancestor's target merge branch.

### PR DESCRIPTION
**Pull request by 6543**
close #11552

### CODE REVIEW

**Comment by blaggacao** [comment]
Failing test seem unrelated, right?

File: routers/api/v1/repo/pull.go [hunk 1]
```diff
@@ -508,6 +510,30 @@ func EditPullRequest(ctx *context.APIContext, form api.EditPullRequestOption) {
...
+		if !ctx.Repo.GitRepo.IsBranchExist(form.Base) {
+			ctx.Error(http.StatusNotFound, "NewBaseBranchNotExist", fmt.Errorf("new base '%s' not exist", form.Base))
+			return
+		}
+		if err := pull_service.ChangeTargetBranch(pr, ctx.User, form.Base); err != nil {
+			if models.IsErrPullRequestAlreadyExists(err) {
+				ctx.Error(http.StatusConflict, "IsErrPullRequestAlreadyExists", err)
+				return
+			} else if models.IsErrIssueIsClosed(err) {
+				ctx.Error(http.StatusUnprocessableEntity, "IsErrPullRequestAlreadyExists", err)
```
Comment (blaggacao) [inline routers/api/v1/repo/pull.go]:
```suggestion
				ctx.Error(http.StatusUnprocessableEntity, "IsErrIssueIsClosed", err)
```

File: routers/api/v1/repo/pull.go [hunk 2]
```diff
@@ -508,6 +510,30 @@ func EditPullRequest(ctx *context.APIContext, form api.EditPullRequestOption) {
...
+				return
+			} else if models.IsErrPullRequestHasMerged(err) {
+				ctx.Error(http.StatusConflict, "IsErrPullRequestAlreadyExists", err)
```
Comment (blaggacao) [inline routers/api/v1/repo/pull.go]:
```suggestion
				ctx.Error(http.StatusConflict, "IsErrPullRequestHasMerged", err)
```

**Review by lafriks — APPROVED** [review]

File: routers/api/v1/repo/pull.go [hunk 3]
```diff
@@ -508,6 +510,30 @@ func EditPullRequest(ctx *context.APIContext, form api.EditPullRequestOption) {
...
+				return
+			} else {
+				ctx.InternalServerError(err)
```
Comment (blaggacao) [inline routers/api/v1/repo/pull.go:612]:
```suggestion
			} else if models.IsErrBranchesEqual(err) {
				ctx.Error(http.StatusUnprocessableEntity, "IsErrBranchesEqual", err)
			} else {
				ctx.InternalServerError(err)
```

**Review by blaggacao — CHANGES_REQUESTED** [review]
Some things I took from: https://github.com/go-gitea/gitea/blob/5cb201dc93bf41556556f2154ea28755907fd550/routers/repo/pull.go#L1190-L1229

File: routers/api/v1/repo/pull.go (hunk 3 above)
Comment (blaggacao) [reply routers/api/v1/repo/pull.go:612]: Not sure about `http.StatusUnprocessableEntity` vs `http.StatusConflict`

File: integrations/api_pull_test.go [hunk 4]
```diff
@@ -90,6 +90,21 @@ func TestAPICreatePullSuccess2(t *testing.T) {
...
+	})
+	resp = session.MakeRequest(t, req, 201)
+	DecodeJSON(t, resp, pull)
+	assert.EqualValues(t, "feature/1", pull.Base.Name)
 
-	session.MakeRequest(t, req, 201)
+	req = NewRequestWithJSON(t, http.MethodPost, fmt.Sprintf("/api/v1/repos/%s/%s/pulls?token=%s", owner10.Name, repo10.Name, token), &api.EditPullRequestOption{
+		Base: "not-exist",
+	})
+	session.MakeRequest(t, req, 404)
```
Comment (blaggacao) [inline integrations/api_pull_test.go:109]: Should it be considered ok _not_ to test:
`IsErrPullRequestAlreadyExists`, `IsErrIssueIsClosed`, `IsErrPullRequestHasMerged` & `IsErrBranchesEqual`?

I could imagine they are sufficiently tested elsewhere, just wanted to make sure, it's ok.

File: routers/api/v1/repo/pull.go (hunk 3 above)
Comment (blaggacao) [reply routers/api/v1/repo/pull.go:612]: Maybe instead of `ctx.InternalServerError` -> `ctx.ServerError("UpdatePullRequestTarget", err)`?

Comment (6543) [reply routers/api/v1/repo/pull.go:612]: @blaggacao IsErrBranchesEqual should not happen since it is checked before:
> && form.Base != pr.BaseBranch

Comment (blaggacao) [reply routers/api/v1/repo/pull.go:612]: ah, yep sorry

**Review by blaggacao — APPROVED** [review]

**Comment by 6543** [comment]
updated tests, now all should PASS :D

**Comment by 6543** [comment]
ready to merge 🚀

**Comment by lunny** [comment]
Is this change compitable with Github API?

**Comment by blaggacao** [comment]
What I know: Github provides this feature through it's API.

**Comment by 6543** [comment]
@lunny https://developer.github.com/v3/pulls/#input-2

**Comment by blaggacao** [comment]
@zeripath @lunny Anything is blocking second review? Are code owners summoned?

File: routers/api/v1/repo/pull.go [hunk 5]
```diff
@@ -508,6 +510,30 @@ func EditPullRequest(ctx *context.APIContext, form api.EditPullRequestOption) {
 		notification.NotifyIssueChangeStatus(ctx.User, issue, statusChangeComment, issue.IsClosed)
 	}
 
+	// change pull target branch
+	if len(form.Base) != 0 && form.Base != pr.BaseBranch {
+		if !ctx.Repo.GitRepo.IsBranchExist(form.Base) {
```
Comment (zeripath) [inline routers/api/v1/repo/pull.go:597]: If we're going to error out we should do it before we make any changes. (This is not your fault - the same problem occurs throughout this function.) But it's kinda crazy that we could change all of the other settings, return an error and still have those changes persisting.)

File: routers/api/v1/repo/pull.go [hunk 6]
```diff
@@ -508,6 +510,30 @@ func EditPullRequest(ctx *context.APIContext, form api.EditPullRequestOption) {
...
+			ctx.Error(http.StatusNotFound, "NewBaseBranchNotExist", fmt.Errorf("new base '%s' not exist", form.Base))
+			return
+		}
+		if err := pull_service.ChangeTargetBranch(pr, ctx.User, form.Base); err != nil {
```
Comment (zeripath) [inline routers/api/v1/repo/pull.go:601]: All of these changes should be within the same transaction.

**Comment by blaggacao** [comment]
https://github.com/git-town/git-town/pull/1518 has gotten merged, git town now supports gitea, but for full support this PR is pressing, everything else is set up. :hourglass: The gun is loaded.

**Comment by blaggacao** [comment]
@zeripath You marked this for 1.13.0. Given it's blocking https://github.com/git-town/git-town/issues/1530, any chance to get this shipped earlier?

**Comment by techknowlogick** [comment]
@blaggacao as we've already passed the feature freeze, and are now in the RC phase for 1.12 we cannot get it into the 1.12 release.

**Comment by blaggacao** [comment]
I see, ok then, let's head for `1.13.0`! :rocket:

**Review by zeripath — APPROVED** [review]
notwithstanding my comments re:transactions as this doesn't make the function worse I think we can leave that be.

**Comment by zeripath** [comment]
@techknowlogick it might be reasonable to backport this as you argue it's a bug that the API doesn't have this when the UI does.

**Comment by zeripath** [comment]
make lg-tm work

**Comment by 6543** [comment]
Backport = #11796

**Comment by blaggacao** [comment]
Nice, thank you all!

**Pull request merged**